pkg/config/reloader.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// ComponentReloader describes how to reload a component.
+type ComponentReloader struct {
+	Name     string
+	Reloader func(*Config) error
+}
+
+// ConfigReloader holds all information required to reload Parca's config into its running components.
+type ConfigReloader struct {
+	logger            log.Logger
+	filename          string
+	watcher           *fsnotify.Watcher
+	reloaders         []ComponentReloader
+	triggerReload     chan struct{}
+	configSuccess     prometheus.Gauge
+	configSuccessTime prometheus.Gauge
+}
+
+// NewConfigReloader returns an instantiated config reloader.
+func NewConfigReloader(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	filename string,
+	reloaders []ComponentReloader,
+) (*ConfigReloader, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to establish config watcher", "err", err)
+		return nil, err
+	}
+
+	if err := watcher.Add(filename); err != nil {
+		level.Error(logger).Log("msg", "failed to start watching config file", "err", err, "path", filename)
+		return nil, err
+	}
+
+	r := &ConfigReloader{
+		logger:   logger,
+		filename: filename,
+		watcher:  watcher,
+
+		reloaders: reloaders,
+
+		triggerReload: make(chan struct{}, 1),
+
+		configSuccess: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "parca_config_last_reload_successful",
+			Help: "Whether the last configuration reload attempt was successful.",
+		}),
+		configSuccessTime: prometheus.NewGauge(prometheus.GaugeOpts{
+			Name: "parca_config_last_reload_success_timestamp_seconds",
+			Help: "Timestamp of the last successful configuration reload.",
+		}),
+	}
+
+	reg.MustRegister(
```

Comment:
Better to use `Register` and propagate the error, rather than panicing.

Reply:
`reg.MustRegister()` is used consistently in Parca, I have no problem changing for `Register()` here, just curious, what makes this one different?

Reply:
IMHO, we should always prefer the methods that return errors over panicking. In this case, the constructor method already returns an error and we handle it on the caller site. If we have usage that fits this type of error handling we should convert them as well.

Reply:
Makes sense, thank you for the explanation